--- Day 23: LAN Party ---
As The Historians wander around a secure area at Easter Bunny HQ, you come across posters for a LAN party scheduled for today! Maybe you can find it; you connect to a nearby datalink port and download a map of the local network (your puzzle input).

The network map provides a list of every connection between two computers. For example:

kh-tc
qp-kh
de-cg
ka-co
yn-aq
qp-ub
cg-tb
vc-aq
tb-ka
wh-tc
yn-cg
kh-ub
ta-co
de-co
tc-td
tb-wq
wh-td
ta-ka
td-qp
aq-cg
wq-ub
ub-vc
de-ta
wq-aq
wq-vc
wh-yn
ka-de
kh-ta
co-tc
wh-qp
tb-vc
td-yn
Each line of text in the network map represents a single connection; the line kh-tc represents a connection between the computer named kh and the computer named tc. Connections aren't directional; tc-kh would mean exactly the same thing.

LAN parties typically involve multiplayer games, so maybe you can locate it by finding groups of connected computers. Start by looking for sets of three computers where each computer in the set is connected to the other two computers.

In this example, there are 12 such sets of three inter-connected computers:

aq,cg,yn
aq,vc,wq
co,de,ka
co,de,ta
co,ka,ta
de,ka,ta
kh,qp,ub
qp,td,wh
tb,vc,wq
tc,td,wh
td,wh,yn
ub,vc,wq
If the Chief Historian is here, and he's at the LAN party, it would be best to know that right away. You're pretty sure his computer's name starts with t, so consider only sets of three computers where at least one computer's name starts with t. That narrows the list down to 7 sets of three inter-connected computers:

co,de,ta
co,ka,ta
de,ka,ta
qp,td,wh
tb,vc,wq
tc,td,wh
td,wh,yn
Find all the sets of three inter-connected computers. How many contain at least one computer with a name that starts with t?

In [101]:
sample_data = """kh-tc
qp-kh
de-cg
ka-co
yn-aq
qp-ub
cg-tb
vc-aq
tb-ka
wh-tc
yn-cg
kh-ub
ta-co
de-co
tc-td
tb-wq
wh-td
ta-ka
td-qp
aq-cg
wq-ub
ub-vc
de-ta
wq-aq
wq-vc
wh-yn
ka-de
kh-ta
co-tc
wh-qp
tb-vc
td-yn"""

In [102]:
sample_data = [line.split('-') for line in sample_data.splitlines()]


In [103]:
from collections import defaultdict
pairwise = defaultdict(set)
for comp1, comp2 in sample_data:
    pairwise[comp1].add(comp2)
    pairwise[comp2].add(comp1)

In [104]:
from itertools import combinations

triples = set()
visited = set()
for c1, cnc in pairwise.items():
    for c2, c3 in combinations(cnc, 2):
        if c2 not in visited and c3 not in visited:
            if c3 in pairwise[c2]:
                triples.add(frozenset([c1, c2, c3]))
    visited.add(c1)

In [105]:
triples

{frozenset({'aq', 'cg', 'yn'}),
 frozenset({'co', 'de', 'ta'}),
 frozenset({'co', 'ka', 'ta'}),
 frozenset({'aq', 'vc', 'wq'}),
 frozenset({'tb', 'vc', 'wq'}),
 frozenset({'tc', 'td', 'wh'}),
 frozenset({'kh', 'qp', 'ub'}),
 frozenset({'td', 'wh', 'yn'}),
 frozenset({'ub', 'vc', 'wq'}),
 frozenset({'de', 'ka', 'ta'}),
 frozenset({'co', 'de', 'ka'}),
 frozenset({'qp', 'td', 'wh'})}

In [106]:
t_triples = set()
visited = set()
for c1, cnc in pairwise.items():
    if c1[0] == 't':
        for c2, c3 in combinations(cnc, 2):
            if c2 not in visited and c3 not in visited:
                if c3 in pairwise[c2]:
                    t_triples.add(frozenset([c1, c2, c3]))
        visited.add(c1)
t_triples

{frozenset({'co', 'ka', 'ta'}),
 frozenset({'co', 'de', 'ta'}),
 frozenset({'tb', 'vc', 'wq'}),
 frozenset({'tc', 'td', 'wh'}),
 frozenset({'td', 'wh', 'yn'}),
 frozenset({'de', 'ka', 'ta'}),
 frozenset({'qp', 'td', 'wh'})}

In [107]:
# Having validated our code with the sample data, let's do the puzzle input:
with open('input23.txt') as file:
    data = [line.rstrip().split('-') for line in file]

In [108]:
pairwise = defaultdict(set)
for comp1, comp2 in data:
    pairwise[comp1].add(comp2)
    pairwise[comp2].add(comp1)

In [109]:
t_triples = set()
visited = set()
for c1, cnc in pairwise.items():
    if c1[0] == 't':
        for c2, c3 in combinations(cnc, 2):
            if c2 not in visited and c3 not in visited:
                if c3 in pairwise[c2]:
                    t_triples.add(frozenset([c1, c2, c3]))
        visited.add(c1)

In [110]:
len(t_triples)

1043

Your puzzle answer was 1043.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
There are still way too many results to go through them all. You'll have to find the LAN party another way and go there yourself.

Since it doesn't seem like any employees are around, you figure they must all be at the LAN party. If that's true, the LAN party will be the largest set of computers that are all connected to each other. That is, for each computer at the LAN party, that computer will have a connection to every other computer at the LAN party.

In the above example, the largest set of computers that are all connected to each other is made up of co, de, ka, and ta. Each computer in this set has a connection to every other computer in the set:

ka-co
ta-co
de-co
ta-ka
de-ta
ka-de
The LAN party posters say that the password to get into the LAN party is the name of every computer at the LAN party, sorted alphabetically, then joined together with commas. (The people running the LAN party are clearly a bunch of nerds.) In this example, the password would be co,de,ka,ta.

What is the password to get into the LAN party?

In [111]:
max_pairs = max(len(coll) for coll in pairwise.values()) + 1
max_pairs

14

In [112]:
len({k: v for k, v in pairwise.items() if len(v) == 13}) # all 520 computers have 13 connections exactly

520

In [ ]:
# since the solution requires a unique largest LAN we can stop early
def find_lan(n):
    visited = set()
    counter = 0
    for c1, cnc in pairwise.items():
        for draw in combinations(cnc, n-1):
            if all(c not in visited for c in draw):
                if all(c3 in pairwise[c2] for c2, c3 in combinations(draw, 2)):
                    return set([*draw, c1])
        visited.add(c1)
find_lan(13)

{'ai', 'bk', 'dc', 'dx', 'fo', 'gx', 'hk', 'kd', 'os', 'uz', 'xn', 'yk', 'zs'}

In [127]:
for n in range(max_pairs, 0, -1):
    if largest_set := find_lan(n):
        break
largest_set

{'ai', 'bk', 'dc', 'dx', 'fo', 'gx', 'hk', 'kd', 'os', 'uz', 'xn', 'yk', 'zs'}

In [128]:
solution = ','.join(sorted(largest_set))
solution

'ai,bk,dc,dx,fo,gx,hk,kd,os,uz,xn,yk,zs'

That's the right answer! You are one gold star closer to finding the Chief Historian.